functions/go/source-gcloud-config/exec/gcloud.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package exec
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var domainRegex = regexp.MustCompile(`\S+@(\S+)`)
+var PullGcloudConfig = PullLocalGcloudConfig
+
+// PullLocalGcloudConfig executes `gcloud` commands to get default gcloud config value.
+func PullLocalGcloudConfig() (map[string]string, error) {
+	projectID := getGcloudConfig("project")
+	if projectID == "" {
+		return nil, fmt.Errorf("`project` has not been set in `gcloud`")
+	}
+	zone := getGcloudConfig("compute/zone")
+	if zone == "" {
+		// Skip if `compute/zone` has not been set in `gcloud`
+	}
+	region := getGcloudConfig("compute/region")
+	if region == "" {
+		// Skip if `compute/region` has not been set in `gcloud`
+	}
+	var domain string
+	if account := getGcloudConfig("core/account"); account == "" {
+		// Skip if `core/account` has not been set in `gcloud`
+	} else {
+		// e.g. account `NAME@COMPANY.com` has matching domain `COMPANY.com`
+		matches := domainRegex.FindStringSubmatch(account)
+		if len(matches) < 2 {
+			// Skip if cannot parse `domain` from gcloud `core/account`
+		} else {
+			domain = matches[1]
+		}
+	}
+	orgID, err := getGcloudOrgID(projectID)
+	if err != nil {
+		return nil, err
+	}
+	if orgID == "" {
+		// Skip if `Organization` or `Folder` not set up.
+	}
+	return map[string]string{
+		"namespace": projectID,
+		"projectID": projectID,
+		"zone":      zone,
+		"region":    region,
+		"domain":    domain,
+		"orgID":     orgID,
+	}, nil
+
+}
+
+func getGcloudConfig(property string) string {
+	var cmdOut, cmdErr bytes.Buffer
+	cmd := exec.Command("gcloud", "config", "get-value", property)
```

Comment:
There may be more performant way to extract the config than via execing gcloud binary, or execing it multiple times.

I suspect the gcloud config file structure is documented and can be directly parsed?

Reply:
> There may be more performant way to extract the config than via execing gcloud binary, or execing it multiple times.
> 
> I suspect the gcloud config file structure is documented and can be directly parsed?

I was exactly thinking the same 😄   But really cannot find where those gcloud configs are stored (searched internal docs and google3). Another option is to use `gcloud config list` . I can go that way if it's better.

Reply:
${HOME}/.config/gcloud/... seems to be where they are. There seems to be some indirection happening but with some luck we can figure it out :)

Reply:
oh, how come I forgot HOME/.config

Reply:
`gcloud config list --format json` could also give parsable JSON output

Reply:
@bharathkkb prefer not using gcloud config list here because the output is in ini file format. It's different than the org-id one.

Reply:
Generally it is [not recommended](https://cloud.google.com/sdk/docs/scripting-gcloud#:~:text=Don%27t%20depend%20on%20the%20raw%20output%20of%20messages%20printed%20to%20standard%20output) to rely on gcloud stdout. I haven't looked into the ini format so it maybe as stable as json op 😅